modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,510 @@
+    }
+
+    /**
+     * Stops all the processes with the passed exception.
+     *
+     * @param err The interrupt reason.
+     */
+    void interrupt(Throwable err) {
+        contexts.forEach((snpName, ctx) -> {
+            if (ctx.fut != null)
+                ctx.fut.onDone(err);
+        });
+
+        contexts.clear();
+
+        clusterOpFuts.forEach((reqId, fut) -> fut.onDone(err));
+    }
+
+    /** Phase 2 and process finish. */
+    private IgniteInternalFuture<?> reduceValidatePartsAndFinish(
+        UUID reqId,
+        Map<UUID, SnapshotCheckResponse> results,
+        Map<UUID, Throwable> errors
+    ) {
+        SnapshotCheckContext ctx = context(null, reqId);
+
+        if (ctx == null)
+            return new GridFinishedFuture<>();
+
+        contexts.remove(ctx.req.snapshotName());
+
+        if (log.isInfoEnabled())
+            log.info("Finished snapshot validation [req=" + ctx.req + ']');
+
+        GridFutureAdapter<SnapshotPartitionsVerifyTaskResult> clusterOpFut = clusterOpFuts.get(reqId);
+
+        if (clusterOpFut == null)
+            return new GridFinishedFuture<>();
+
+        assert results.values().stream().noneMatch(res -> res != null && res.metas != null);
+
+        if (ctx.req.allRestoreHandlers) {
+            try {
+                if (!F.isEmpty(errors))
+                    throw F.firstValue(errors);
+
+                Map<ClusterNode, Map<String, SnapshotHandlerResult<?>>> cstRes = mapCustomHandlersResults(results, ctx.req.nodes());
+
+                kctx.cache().context().snapshotMgr().checker().checkCustomHandlersResults(ctx.req.snapshotName(), cstRes);
+
+                clusterOpFut.onDone(new SnapshotPartitionsVerifyTaskResult(ctx.clusterMetas, null));
+            }
+            catch (Throwable err) {
+                clusterOpFut.onDone(err);
+            }
+        }
+        else {
+            Map<ClusterNode, Exception> errors0 = mapErrors(errors);
+
+            if (!F.isEmpty(results)) {
+                Map<ClusterNode, Map<PartitionKeyV2, PartitionHashRecordV2>> results0 = mapPartsHashes(results, ctx.req.nodes());
+
+                IdleVerifyResultV2 chkRes = SnapshotChecker.reduceHashesResults(results0, errors0);
+
+                clusterOpFut.onDone(new SnapshotPartitionsVerifyTaskResult(ctx.clusterMetas, chkRes));
+            }
+            else {
+                assert !errors0.isEmpty();
+
+                clusterOpFut.onDone(new IgniteSnapshotVerifyException(errors0));
+            }
+        }
+
+        return new GridFinishedFuture<>();
+    }
+
+    /** Phase 2 beginning.  */
+    private IgniteInternalFuture<SnapshotCheckResponse> validateParts(SnapshotCheckProcessRequest req) {
+        SnapshotCheckContext ctx;
+
+        // The context can be null, if a required node leaves before this phase.
+        if (!req.nodes().contains(kctx.localNodeId()) || (ctx = context(null, req.requestId())) == null || ctx.locMeta == null)
```

Comment:
does `ctx.locMeta == null` cover case `!req.nodes().contains(kctx.localNodeId())`?

Reply:
No. We've added client-initiator to the required nodes. It has no meta. Previously, the required nodes were only data nodes. Also snapshot might be placed in any manner to the baseline nodes and/or be restored from another cluster. Any node may miss snapshot meta. Test `testRestoreFromAnEmptyNode()` show case like that.